EIPS/eip-7484.md

```diff
@@ -0,0 +1,217 @@
+
+However, there are obviously tradeoffs for using a singleton. A Singleton Registry creates a single point of failure that, if exploited, could lead to serious consequences for smart accounts. The most serious attack vector of these would be the ability for an attacker to attest to a malicious module on behalf of a trusted attester. One tradeoff here is that using multiple registries, changes in security attestations (for example a vulnerability is found and an attestation is revoked) are slower to propagate across the ecosystem, giving attackers an opportunity to exploit vulnerabilities for longer or even find and exploit them after seeing an issue pointed out in a specific Registry but not in others.
+
+Due to being a singleton, the Registry needs to be very flexible and thus likely less computationally efficient in comparison to a narrow, optimised Registry. This means that querying a Singleton Registry is likely to be more computationally (and by extension gas) intensive than querying a more narrow Registry. The tradeoff here is that a singleton makes it cheaper to query attestations from multiple parties simultaneously. So, depending on the Registry architectures, there is an amount of attestations to query (N) after which using a flexible singleton is actually computationally cheaper than querying N narrow registries. However, the reference implementation has also been designed with gas usage in mind and it is unlikely that specialised registries will be able to significantly decrease gas beyond the reference implementations benchmarks.
+
+### Adapter
+
+In order for smart accounts to increase security guarantees when adding modules, they must be able to securely query the Module Registry and handle the return data correctly. In order to achieve this, this proposal aims to provide a standardised interface that may be implemented by smart accounts irrespective of their architecture, execution flows and security assumptions.
+
+### Related work
+
+The reference implementation of the Registry is heavily inspired by the Ethereum Attestation Service. The specific use-case of this proposal, however, required some custom modifications and additions to EAS, meaning that using the existing EAS contracts as the Module Registry was sub-optimal.
+
+## Backwards Compatibility
+
+No backward compatibility issues found.
+
+## Reference Implementation
+
+### Adapter.sol
+
+```solidity
+contract Adapter {
+    IRegistry registry;
+
+    function checkModule(address module, address trustedAttester) internal {
+       // check module implementation address on registry
+        (uint48 listedAt, uint48 flaggedAt) = registry.check(module, trustedAttester);
+
+        // revert if module was ever flagged or was never attested to
+        require(listedAt != 0 && flaggedAt == 0, "Module is insecure");
+    }
+
+    function verifyModule(address module, address[] memory attesters, uint256 threshold) internal {
```

Comment:
Based on the interface defined here, it looks like checks are made based on whether an Attestation exists or not, i.e. an attestation can be considered to be a boolean. 
What are your thoughts on supporting checks based on the data inside an Attestation? Say, for example, an Attestation stores the number of audits done for a particular module and a smart contract wallet wishes to set a security policy where only modules with >= K audits are installed.